src/RoslynPad.Avalonia/Folding/CShapRoslynFoldingStrategy.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System.Text.RegularExpressions;
+using AvaloniaEdit.Folding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using TextDocument = AvaloniaEdit.Document.TextDocument;
+
+namespace RoslynPad.Folding;
+
+/// <summary>
+/// Folding strategy based on <see cref="Microsoft.CodeAnalysis.CSharp.FoldingStrategy"></see> Roslyn lexer
+/// </summary>
+public partial class CShapRoslynFoldingStrategy : FoldingStrategy
+{
+    readonly List<NewFolding> _foldings = [];
+    readonly string _defaultFoldingText = "...";
+    bool _startFolded;
+
+    /// <summary>
+    /// Folding SyntaxKind types <see cref="Microsoft.CodeAnalysis.CSharp.SyntaxKind"></see>
+    /// </summary>
+    readonly Dictionary<SyntaxKind, SyntaxKind> _pairTypes = new()
+    {
+        {SyntaxKind.OpenBraceToken,  SyntaxKind.CloseBraceToken},
+        {SyntaxKind.OpenBracketToken,  SyntaxKind.CloseBracketToken},
+        {SyntaxKind.OpenParenToken,  SyntaxKind.CloseParenToken},
+        {SyntaxKind.RegionDirectiveTrivia,  SyntaxKind.EndRegionDirectiveTrivia},
+        {SyntaxKind.MultiLineCommentTrivia, SyntaxKind.MultiLineCommentTrivia},
+        {SyntaxKind.SingleLineDocumentationCommentTrivia, SyntaxKind.SingleLineDocumentationCommentTrivia},
+        {SyntaxKind.UsingDirective, SyntaxKind.UsingDirective}
+    };
+
+    protected override IEnumerable<NewFolding> CreateNewFoldings(TextDocument document, out int firstErrorOffset)
+    {
+        firstErrorOffset = -1;
+
+        _foldings.Clear();
+
+        _startFolded = false;
+        var tree = CSharpSyntaxTree.ParseText(document.Text);
```

Comment:
BTW there's no need to reparse the document, Roslyn is already doing that in the background continuously

Reply:
Fixed